### Problem Statement
How to optimize sales and cut down on costs due to the economic down turn

### Problem Framing



### Steps to optimize sales and reduce cost
To  optimize sales , the following can be done:
- Indentify target audience : This is needed to undertand your ideal customer to tailor marketing efforts.
- Develop a unique value proposition : Clearly communicate your products benefits.
- Set sales goals and metrics : Establish targets and track progress.
- Enhance sales channels : Optimize websites , social media and in- store experiences.
- Leverage data analtyics : Use insights to refine sales strategies

To reduce costs,the following has to be noted:

- Analyse expense : Identity areas for cost reduction.
- Implement cost-saving technologies : Automate processes and enhance efficiency.
- Optimize inventory management : Minimize waste and excess stock
- Streamline operations: Eliminate unnecessary proceses and reduce waste

##### Most importantly use data-driven decision making to inform sales and cost reductions strategies so as to improve overall business performance.

###### Steps needed to be done the the data.
- Data exploration
- Data Cleaning
- Getting insights from Data


In [1]:
# Import libraries 
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sb
%matplotlib inline


In [2]:
data = pd.read_csv('retail.csv',encoding='unicode_escape')
data.head()

C:\Users\USER\AppData\Local\Temp\ipykernel_10156\3995310157.py:1: DtypeWarning: Columns (5) have mixed types. Specify dtype option on import or set low_memory=False.
  data = pd.read_csv('retail.csv',encoding='unicode_escape')


Invoice StockCode                          Description  Quantity  \
0  489434     85048  15CM CHRISTMAS GLASS BALL 20 LIGHTS        12   
1  489434    79323P                   PINK CHERRY LIGHTS        12   
2  489434    79323W                  WHITE CHERRY LIGHTS        12   
3  489434     22041         RECORD FRAME 7" SINGLE SIZE         48   
4  489434     21232       STRAWBERRY CERAMIC TRINKET BOX        24   

        InvoiceDate  Price  Customer ID Country  
0  12/01/2009 07:45  $6.95      13085.0      UK  
1  12/01/2009 07:45  $6.75      13085.0      UK  
2  12/01/2009 07:45   6.75      13085.0      UK  
3  12/01/2009 07:45    2.1      13085.0      UK  
4  12/01/2009 07:45   1.25      13085.0      UK

### Data Cleaning Processes


In [4]:
#Check Country Column
data['Country'].value_counts()

Country
United Kingdom          485830
EIRE                      9670
Germany                   8129
France                    5772
Netherlands               2769
Spain                     1278
Switzerland               1187
Portugal                  1101
Belgium                   1054
Channel Islands            906
Sweden                     902
Italy                      731
Australia                  654
Cyprus                     554
Austria                    537
Greece                     517
United Arab Emirates       432
Denmark                    428
Norway                     369
Finland                    354
Unspecified                310
USA                        244
Japan                      224
Poland                     194
Malta                      172
Lithuania                  154
Singapore                  117
RSA                        111
Bahrain                    107
Canada                      77
Hong Kong                   76
Thailand                    76


##### Change Countries to full names e.g UK to United Kingdom, USA to United States of America

In [5]:
data['Country'] = data['Country'].replace({'UK':'United Kingdom','EIRE':'Ireland','USA':'United States of America','RSA':'South Africa'})

In [6]:
# Check
data['Country'].value_counts()

Country
United Kingdom              485852
Ireland                       9670
Germany                       8129
France                        5772
Netherlands                   2769
Spain                         1278
Switzerland                   1187
Portugal                      1101
Belgium                       1054
Channel Islands                906
Sweden                         902
Italy                          731
Australia                      654
Cyprus                         554
Austria                        537
Greece                         517
United Arab Emirates           432
Denmark                        428
Norway                         369
Finland                        354
Unspecified                    310
United States of America       244
Japan                          224
Poland                         194
Malta                          172
Lithuania                      154
Singapore                      117
South Africa                   111
Bahrain     

In [7]:
# Check Description
data['Description'].value_counts()

Description
WHITE HANGING HEART T-LIGHT HOLDER     3549
REGENCY CAKESTAND 3 TIER               2212
STRAWBERRY CERAMIC TRINKET BOX         1843
PACK OF 72 RETRO SPOT CAKE CASES       1466
ASSORTED COLOUR BIRD ORNAMENT          1457
                                       ... 
stock credited from royal yacht inc       1
VINTAGE METAL CAKE STAND CREAM            1
BLUE BAROQUE FLOCK CANDLE HOLDER          1
S/4 HEART CRYSTAL FRIDGE MAGNETS          1
dotcom email                              1
Name: count, Length: 4681, dtype: int64

In [8]:
# Change all to Caps
data['Description'] = data['Description'].str.upper()

In [9]:
# check
data['Description']

0         15CM CHRISTMAS GLASS BALL 20 LIGHTS
1                          PINK CHERRY LIGHTS
2                         WHITE CHERRY LIGHTS
3                RECORD FRAME 7" SINGLE SIZE 
4              STRAWBERRY CERAMIC TRINKET BOX
                         ...                 
525456                   FELTCRAFT DOLL ROSIE
525457           FELTCRAFT PRINCESS LOLA DOLL
525458         FELTCRAFT PRINCESS OLIVIA DOLL
525459     PINK FLORAL FELTCRAFT SHOULDER BAG
525460                 JUMBO STORAGE BAG SUKI
Name: Description, Length: 525461, dtype: object

In [10]:
#check Customer ID
data['Customer ID']

0         13085.0
1         13085.0
2         13085.0
3         13085.0
4         13085.0
           ...   
525456    17530.0
525457    17530.0
525458    17530.0
525459    17530.0
525460    17530.0
Name: Customer ID, Length: 525461, dtype: float64

In [11]:
# To remove the full stop and last 0
# First Convert to String 
data['Customer ID'] = data['Customer ID'].astype('str')

In [12]:
# Replace all full stop with nothing
data['Customer ID'] = data['Customer ID'].str.replace('.','')

In [13]:
#remove last zero
data['Customer ID'] = data['Customer ID'].str[:5]

In [14]:
#check
data['Customer ID']

0         13085
1         13085
2         13085
3         13085
4         13085
          ...  
525456    17530
525457    17530
525458    17530
525459    17530
525460    17530
Name: Customer ID, Length: 525461, dtype: object

#### Convert InvoiceDate Column to DateTime DataType

In [15]:
data['InvoiceDate'] = pd.to_datetime(data['InvoiceDate'])

In [16]:
#Check
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 525461 entries, 0 to 525460
Data columns (total 8 columns):
 #   Column       Non-Null Count   Dtype         
---  ------       --------------   -----         
 0   Invoice      525461 non-null  object        
 1   StockCode    525461 non-null  object        
 2   Description  522533 non-null  object        
 3   Quantity     525461 non-null  int64         
 4   InvoiceDate  525461 non-null  datetime64[ns]
 5   Price        525461 non-null  object        
 6   Customer ID  525461 non-null  object        
 7   Country      525461 non-null  object        
dtypes: datetime64[ns](1), int64(1), object(6)
memory usage: 32.1+ MB


#### To remove last alphabet from the invoice number

In [17]:
data['StockCode']

0          85048
1         79323P
2         79323W
3          22041
4          21232
           ...  
525456     22271
525457     22750
525458     22751
525459     20970
525460     21931
Name: StockCode, Length: 525461, dtype: object

In [18]:
data['StockCode'] = data['StockCode'].str[:5]

In [19]:
#Check
data['StockCode']

0         85048
1         79323
2         79323
3         22041
4         21232
          ...  
525456    22271
525457    22750
525458    22751
525459    20970
525460    21931
Name: StockCode, Length: 525461, dtype: object

In [20]:
#Check Data again
data.head()

Invoice StockCode                          Description  Quantity  \
0  489434     85048  15CM CHRISTMAS GLASS BALL 20 LIGHTS        12   
1  489434     79323                   PINK CHERRY LIGHTS        12   
2  489434     79323                  WHITE CHERRY LIGHTS        12   
3  489434     22041         RECORD FRAME 7" SINGLE SIZE         48   
4  489434     21232       STRAWBERRY CERAMIC TRINKET BOX        24   

          InvoiceDate  Price Customer ID         Country  
0 2009-12-01 07:45:00  $6.95       13085  United Kingdom  
1 2009-12-01 07:45:00  $6.75       13085  United Kingdom  
2 2009-12-01 07:45:00   6.75       13085  United Kingdom  
3 2009-12-01 07:45:00    2.1       13085  United Kingdom  
4 2009-12-01 07:45:00   1.25       13085  United Kingdom

data.dropna(axis = 0, ) #delete null values on the rows


In [28]:
data.duplicated().sum()

6978

In [29]:
data.duplicated()

0         False
1         False
2         False
3         False
4         False
          ...  
525456    False
525457    False
525458    False
525459    False
525460    False
Length: 525461, dtype: bool

In [31]:
data.duplicated().value_counts()

False    518483
True       6978
Name: count, dtype: int64

In [36]:
data[data.duplicated()].head(30)

Invoice StockCode                         Description  Quantity  \
371   489517     21912            VINTAGE SNAKES & LADDERS         1   
383   489517     22130    PARTY CONE CHRISTMAS DECORATION          6   
384   489517     22319   HAIRCLIPS FORTIES FABRIC ASSORTED        12   
385   489517     21913      VINTAGE SEASIDE JIGSAW PUZZLES         1   
386   489517     21821    GLITTER STAR GARLAND WITH BELLS          1   
390   489517     84951     S/4 PISTACHIO LOVEBIRD COASTERS         1   
391   489517     21491     SET OF THREE VINTAGE GIFT WRAPS         1   
394   489517     21912            VINTAGE SNAKES & LADDERS         1   
657   489529     22028        PENNY FARTHING BIRTHDAY CARD        12   
658   489529     22036              DINOSAUR BIRTHDAY CARD        12   
669   489529     48129                    DOOR MAT TOPIARY         1   
723   489533     22073              RETRO SPOT STORAGE JAR         1   
725   489533     84946     ANTIQUE SILVER TEA GLASS ETCHED         6   
745   489536     21809    CHRISTMAS HANGING TREE WITH BELL         1   
799   489536     22142        CHRISTMAS CRAFT WHITE FAIRY          1   
808   489536     17164    ASS COL SMALL SAND FROG P'WEIGHT         1   
809   489536     22353  LUNCHBOX WITH CUTLERY FAIRY CAKES          1   
810   489536     21791  VINTAGE HEADS AND TAILS CARD GAME          2   
817   489536     22142        CHRISTMAS CRAFT WHITE FAIRY          1   
820   489536     21650      ASSORTED TUTTI FRUTTI BRACELET         1   
840   489536     21231      SWEETHEART CERAMIC TRINKET BOX         1   
849   489536     85033      SET/6 SILVER REINDEER T-LIGHTS         1   
850   489536     22185          SLATE TILE NATURAL HANGING         2   
852   489536     21579     LOLITA  DESIGN  COTTON TOTE BAG         1   
855   489536     21579     LOLITA  DESIGN  COTTON TOTE BAG         1   
856   489536     21034        REX CASH+CARRY JUMBO SHOPPER         1   
863   489536     22107                  PIZZA PLATE IN BOX         1   
865   489536     21786             RAIN HAT WITH RED SPOTS         1   
1073  489548     22196        SMALL HEART MEASURING SPOONS         1   
1082  489548     85053          FRENCH ENAMEL CANDLEHOLDER         1   

             InvoiceDate Price Customer ID         Country  
371  2009-12-01 11:34:00  3.75       16329  United Kingdom  
383  2009-12-01 11:34:00  0.85       16329  United Kingdom  
384  2009-12-01 11:34:00  0.65       16329  United Kingdom  
385  2009-12-01 11:34:00  3.75       16329  United Kingdom  
386  2009-12-01 11:34:00  3.75       16329  United Kingdom  
390  2009-12-01 11:34:00  2.55       16329  United Kingdom  
391  2009-12-01 11:34:00  1.95       16329  United Kingdom  
394  2009-12-01 11:34:00  3.75       16329  United Kingdom  
657  2009-12-01 11:51:00  0.42       17984  United Kingdom  
658  2009-12-01 11:51:00  0.42       17984  United Kingdom  
669  2009-12-01 11:51:00  6.75       17984  United Kingdom  
723  2009-12-01 12:02:00  3.75       16011  United Kingdom  
725  2009-12-01 12:02:00  1.25       16011  United Kingdom  
745  2009-12-01 12:13:00  1.25       16393  United Kingdom  
799  2009-12-01 12:13:00  1.45       16393  United Kingdom  
808  2009-12-01 12:13:00  0.42       16393  United Kingdom  
809  2009-12-01 12:13:00  2.55       16393  United Kingdom  
810  2009-12-01 12:13:00  1.25       16393  United Kingdom  
817  2009-12-01 12:13:00  1.45       16393  United Kingdom  
820  2009-12-01 12:13:00  0.65       16393  United Kingdom  
840  2009-12-01 12:13:00  1.25       16393  United Kingdom  
849  2009-12-01 12:13:00  1.95       16393  United Kingdom  
850  2009-12-01 12:13:00  1.65       16393  United Kingdom  
852  2009-12-01 12:13:00  2.25       16393  United Kingdom  
855  2009-12-01 12:13:00  2.25       16393  United Kingdom  
856  2009-12-01 12:13:00  0.95       16393  United Kingdom  
863  2009-12-01 12:13:00  3.75       16393  United Kingdom  
865  2009-12-01 12:13:00  0.42       16393  United Kingdom  
107

##### Remove dollars sign from price


In [39]:
data.head()


Invoice StockCode                          Description  Quantity  \
0  489434     85048  15CM CHRISTMAS GLASS BALL 20 LIGHTS        12   
1  489434     79323                   PINK CHERRY LIGHTS        12   
2  489434     79323                  WHITE CHERRY LIGHTS        12   
3  489434     22041         RECORD FRAME 7" SINGLE SIZE         48   
4  489434     21232       STRAWBERRY CERAMIC TRINKET BOX        24   

          InvoiceDate  Price Customer ID         Country  
0 2009-12-01 07:45:00  $6.95       13085  United Kingdom  
1 2009-12-01 07:45:00  $6.75       13085  United Kingdom  
2 2009-12-01 07:45:00   6.75       13085  United Kingdom  
3 2009-12-01 07:45:00    2.1       13085  United Kingdom  
4 2009-12-01 07:45:00   1.25       13085  United Kingdom

In [40]:
data['Price'] =  data['Price'] . replace('[\$,]','', regex=True)

In [41]:
data.head()

Invoice StockCode                          Description  Quantity  \
0  489434     85048  15CM CHRISTMAS GLASS BALL 20 LIGHTS        12   
1  489434     79323                   PINK CHERRY LIGHTS        12   
2  489434     79323                  WHITE CHERRY LIGHTS        12   
3  489434     22041         RECORD FRAME 7" SINGLE SIZE         48   
4  489434     21232       STRAWBERRY CERAMIC TRINKET BOX        24   

          InvoiceDate Price Customer ID         Country  
0 2009-12-01 07:45:00  6.95       13085  United Kingdom  
1 2009-12-01 07:45:00  6.75       13085  United Kingdom  
2 2009-12-01 07:45:00  6.75       13085  United Kingdom  
3 2009-12-01 07:45:00   2.1       13085  United Kingdom  
4 2009-12-01 07:45:00  1.25       13085  United Kingdom

In [42]:
data['StockCode'].nunique() 

3581

In [43]:
data['Customer ID'].nunique()

4384

In [44]:
data['Country'].nunique()

40

In [47]:
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 525461 entries, 0 to 525460
Data columns (total 8 columns):
 #   Column       Non-Null Count   Dtype         
---  ------       --------------   -----         
 0   Invoice      525461 non-null  object        
 1   StockCode    525461 non-null  object        
 2   Description  522533 non-null  object        
 3   Quantity     525461 non-null  int64         
 4   InvoiceDate  525461 non-null  datetime64[ns]
 5   Price        525461 non-null  object        
 6   Customer ID  525461 non-null  object        
 7   Country      525461 non-null  object        
dtypes: datetime64[ns](1), int64(1), object(6)
memory usage: 32.1+ MB


In [50]:
data['Price']= data['Price'].astype('float64')

In [51]:
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 525461 entries, 0 to 525460
Data columns (total 8 columns):
 #   Column       Non-Null Count   Dtype         
---  ------       --------------   -----         
 0   Invoice      525461 non-null  object        
 1   StockCode    525461 non-null  object        
 2   Description  522533 non-null  object        
 3   Quantity     525461 non-null  int64         
 4   InvoiceDate  525461 non-null  datetime64[ns]
 5   Price        525461 non-null  float64       
 6   Customer ID  525461 non-null  object        
 7   Country      525461 non-null  object        
dtypes: datetime64[ns](1), float64(1), int64(1), object(5)
memory usage: 32.1+ MB


In [52]:
data.describe()

Quantity                    InvoiceDate          Price
count  525461.000000                         525461  525461.000000
mean       10.337667  2010-06-28 11:37:36.845017856       4.688834
min     -9600.000000            2009-12-01 07:45:00  -53594.360000
25%         1.000000            2010-03-21 12:20:00       1.250000
50%         3.000000            2010-07-06 09:51:00       2.100000
75%        10.000000            2010-10-15 12:45:00       4.210000
max     19152.000000            2010-12-09 20:01:00   25111.090000
std       107.424110                            NaN     146.126914

##### This dataset contains sales record between 1st of December 2009 and 21st of March 2010 with -
- The total number of 525461 quantities sold all together
- The average price of $4.69 per stock
 

In [53]:
data.head()

Invoice StockCode                          Description  Quantity  \
0  489434     85048  15CM CHRISTMAS GLASS BALL 20 LIGHTS        12   
1  489434     79323                   PINK CHERRY LIGHTS        12   
2  489434     79323                  WHITE CHERRY LIGHTS        12   
3  489434     22041         RECORD FRAME 7" SINGLE SIZE         48   
4  489434     21232       STRAWBERRY CERAMIC TRINKET BOX        24   

          InvoiceDate  Price Customer ID         Country  
0 2009-12-01 07:45:00   6.95       13085  United Kingdom  
1 2009-12-01 07:45:00   6.75       13085  United Kingdom  
2 2009-12-01 07:45:00   6.75       13085  United Kingdom  
3 2009-12-01 07:45:00   2.10       13085  United Kingdom  
4 2009-12-01 07:45:00   1.25       13085  United Kingdom

##### Total amount on each stock is quantity * price

In [54]:
data['Total Amount'] = data['Quantity'] * data['Price']

In [55]:
data['Total Amount']

0          83.40
1          81.00
2          81.00
3         100.80
4          30.00
           ...  
525456      5.90
525457      3.75
525458      3.75
525459      7.50
525460      3.90
Name: Total Amount, Length: 525461, dtype: float64

##### Total amount by Invoice 


In [65]:
Amount_by_Invoice = data.groupby(by='Invoice')[['Total Amount']].sum()

In [66]:
Amount_by_Invoice

Total Amount
Invoice              
489434         505.30
489435         145.80
489436         630.33
489437         310.75
489438        2286.24
...               ...
C538121        -12.75
C538122         -1.25
C538123         -7.50
C538124        -17.70
C538164         -1.95

[28816 rows x 1 columns]

##### Quantity and Total Sales per Country

In [61]:
QT_per_Country = data.groupby(by='Country')[['Total Amount','Quantity']].sum()

In [62]:
QT_per_Country

Total Amount  Quantity
Country                                         
Australia                    30051.800     20053
Austria                      13023.280      6479
Bahrain                       2313.150      1015
Belgium                      22663.530     11980
Bermuda                       1253.140      2798
Brazil                         268.270       189
Canada                        1216.660       894
Channel Islands              21367.320     10994
Cyprus                       11216.450      4371
Denmark                      46972.950    227030
Finland                       7187.710      3651
France                      130769.900     74471
Germany                     196290.351    107133
Greece                       14284.970      6151
Hong Kong                     3922.260      2306
Iceland                       1323.320       828
Ireland                     352242.730    188704
Israel                        3199.400      1132
Italy                        13788.840      7310
Japan                         8435.960      6604
Korea                          949.820       598
Lebanon                        172.030        71
Lithuania                     4892.680      2306
Malta                         2686.750      1547
Netherlands                 263863.410    181823
Nigeria                         -6.660        56
Norway                        4079.570      7863
Poland                        3314.950      1991
Portugal                     23769.780     11878
Singapore                     4037.770      1753
South Africa                  2366.720      1618
Spain                        37084.900     18332
Sweden                       51213.510     52238
Switzerland                  43343.410     22053
Thailand                      3070.540      2552
United Arab Emirates          8046.370      5746
United Kingdom             8194777.533   4429046
United States of America      4555.620      2666
Unspecified                   4937.530      3416
West Indies                    536.410       395

In [67]:
QT_per_Country_sorted = QT_per_Country.sort_values(by='Quantity', ascending=False)

In [68]:
QT_per_Country_sorted

Total Amount  Quantity
Country                                         
United Kingdom             8194777.533   4429046
Denmark                      46972.950    227030
Ireland                     352242.730    188704
Netherlands                 263863.410    181823
Germany                     196290.351    107133
France                      130769.900     74471
Sweden                       51213.510     52238
Switzerland                  43343.410     22053
Australia                    30051.800     20053
Spain                        37084.900     18332
Belgium                      22663.530     11980
Portugal                     23769.780     11878
Channel Islands              21367.320     10994
Norway                        4079.570      7863
Italy                        13788.840      7310
Japan                         8435.960      6604
Austria                      13023.280      6479
Greece                       14284.970      6151
United Arab Emirates          8046.370      5746
Cyprus                       11216.450      4371
Finland                       7187.710      3651
Unspecified                   4937.530      3416
Bermuda                       1253.140      2798
United States of America      4555.620      2666
Thailand                      3070.540      2552
Lithuania                     4892.680      2306
Hong Kong                     3922.260      2306
Poland                        3314.950      1991
Singapore                     4037.770      1753
South Africa                  2366.720      1618
Malta                         2686.750      1547
Israel                        3199.400      1132
Bahrain                       2313.150      1015
Canada                        1216.660       894
Iceland                       1323.320       828
Korea                          949.820       598
West Indies                    536.410       395
Brazil                         268.270       189
Lebanon                        172.030        71
Nigeria                         -6.660        56

###### From the above, we have United Kingdom with total sales of $8194777.533 and 4429046 quantities of stock sold

##### Recomendation to stakeholders

Here are some recommendations to shareholders on how to optimise sales and reduce cost
  - Expand or develop new products so as to cater for evolving customer needs
  - Introduction of discount on products so buyers can pay less and buy more.


In [69]:
data.to_csv('Retail_Store_done.csv')